explorer/explorer_ios.go

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: Unlicense OR MIT
+
+//go:build ios
+// +build ios
+
+package explorer
+
+/*
+#cgo CFLAGS: -Werror -xobjective-c -fmodules -fobjc-arc
+
+#include <UIKit/UIKit.h>
+#include <stdint.h>
+
+@interface explorer_delegate:NSObject
+@end
+
+@interface explorer:NSObject<UIDocumentPickerDelegate>
+@property uint64_t mode;
+@property (strong) UIDocumentPickerViewController * picker;
+@property (strong) UIViewController * controller;
+@property uint32_t id;
+- (bool) show:(char*)text;
+- (bool) exportFile:(char*)name;
+- (bool) importFile:(char*)ext;
+@end
+
+static const uint64_t IMPORT_MODE = 1;
+static const uint64_t EXPORT_MODE = 2;
+
+static CFTypeRef createExplorer(uint64_t mode, CFTypeRef controllerRef, int32_t id) {
+	explorer * expl = [[explorer alloc] init];
+	expl.mode = mode;
+	expl.controller = (__bridge UIViewController *)controllerRef;
+	expl.id = id;
+	return (__bridge_retained CFTypeRef)expl;
+}
+
+static bool show(CFTypeRef expl, char * text) {
+	return [(__bridge explorer *)expl show:text];
+}
+
+static CFTypeRef fileWriteHandler(CFTypeRef u) {
+	NSError *err = nil;
+	NSFileHandle *handler = [NSFileHandle fileHandleForWritingToURL:(__bridge NSURL *)u error:&err];
+	if (err != nil) {
+		return 0;
+	}
+	return (__bridge_retained CFTypeRef)handler;
+}
+
+static CFTypeRef fileReadHandler(CFTypeRef u) {
+	NSError *err = nil;
+	NSFileHandle *handler = [NSFileHandle fileHandleForReadingFromURL:(__bridge NSURL *)u error:&err];
+	if (err != nil) {
+		return 0;
+	}
+	return (__bridge_retained CFTypeRef)handler;
+}
+
+static void closeFile(CFTypeRef handler, CFTypeRef u) {
+	[(__bridge NSURL *)u stopAccessingSecurityScopedResource];
+	[(__bridge NSFileHandle *)handler closeFile];
+}
+
+static void fileWrite(CFTypeRef handler, uint8_t* b, uint64_t len) {
+	NSData *data = [NSData dataWithBytes:b length:len];
+
+	NSFileHandle * h = (__bridge NSFileHandle *)handler;
+	[h writeData:data];
+}
+
+static uint64_t fileRead(CFTypeRef handler, uint8_t* b, uint64_t len) {
+	if (@available(iOS 14, *)) {
+		NSError *err = nil;
+
+		NSFileHandle * h = (__bridge NSFileHandle *)handler;
+		NSData *data = [h readDataUpToLength:len error:&err];
+		[data getBytes:b length:data.length];
+
+		return data.length;
+	}
+	return 0;
+}
```

Comment:
Much like the macOS code, I'd love to find a way to put most of the Objective-C/C code into a single file. Splitting the implementation like this makes it harder to understand the control flow. If we do find a viable way to consolidate on macOS, please also apply it here to the extent that it is possible to do so.

Reply:
I fix it on the latest commit (`explorer: [ios] move functions to .m file`]).